File: 594.cpp
/*
We define a harmonious array is an array where the difference between its maximum value and its minimum value is exactly 1.

Now, given an integer array, you need to find the length of its longest harmonious subsequence among all its possible subsequences.

Example 1:
Input: [1,3,2,2,5,2,3,7]
Output: 5
Explanation: The longest harmonious subsequence is [3,2,2,2,3].
Note: The length of the input array will not exceed 20,000.
*/

class Solution {
public:
    
    int findLHS(vector<int>& nums) {
        map<int, int> hashmap;
        for (int num : nums)
            hashmap[num]++;
        
        int r = 0;
        for (auto it = hashmap.begin(); it != hashmap.end(); it++) {
            auto nit = next(it);
            if (nit == hashmap.end()) break;
            if (it->first + 1 != nit->first) continue;
            r = max(r, it->second + nit->second);
        }
        return r;
    }
};

File: 594.java
/*
We define a harmounious array as an array where the difference between its maximum value and its minimum value is exactly 1.

Now, given an integer array, you need to find the length of its longest harmonious subsequence among all its possible subsequences.

Example 1:

Input: [1,3,2,2,5,2,3,7]
Output: 5
Explanation: The longest harmonious subsequence is [3,2,2,2,3].
*/

class Solution {
    public int findLHS(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        int ans = 0;
        for (int num : nums) {
            int v = map.getOrDefault(num, 0);
            map.put(num, v+1);
            ans = Math.max(ans, Math.max(map.getOrDefault(num-1, -(v+1)), map.getOrDefault(num+1, -(v+1))) + v+1);
        }
        return ans;
    }
}
